File: js/quicksort.js
"use strict";

function quickSort(array, start = 0, end = array.length - 1) {
  if (start < end) {
    const pivot_index = partition(array, start, end);
    quickSort(array, start, pivot_index - 1);
    quickSort(array, pivot_index + 1, end);
  }

  return array;
}

function partition(array, start, end) {
  const pivot = array[end];
  let i = start - 1;

  for (let j = start; j < end; j++) {
    if (array[j] <= pivot) {
      i++;
      [array[i], array[j]] = [array[j], array[i]];
    }
  }
  const pivot_index = i + 1;
  [array[pivot_index], array[end]] = [array[end], array[pivot_index]];

  return pivot_index;
}

let array = [4, 6, 1, 2, 1, 6, 3, 3, 3, 6, 26, 1, 2, 19];
console.log(quickSort(array));

File: quicksort.rb
# frozen_string_literal: true

def quicksort(arr, low = 0, high = arr.length - 1)
  if low < high
    pivot_index = arrange(arr, low, high)
    quicksort(arr, low, pivot_index - 1)
    quicksort(arr, pivot_index + 1, high)
  end

  arr
end

def arrange(arr, low, high) # rubocop : disable Metrics/MethodLength
  pivot = arr[high]
  i = low - 1

  (low..high - 1).each do |j|
    if arr[j] > pivot
      i += 1
      arr[j], arr[i] = arr[i], arr[j]
    end
  end
  pivot_index = i + 1
  arr[high], arr[pivot_index] = arr[pivot_index], arr[high]

  pivot_index
end

puts quicksort([3, 5, 1, 1, 60, 3, 2, 1, 1, 5, 6, 54]).join(',')
